File: algorithms/go/quicksort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"time"
)

func partition(arr []int, l, r int) int {
	pivot, i := arr[r], l
	for j := l; j < r; j++ {
		if arr[j] <= pivot {
			arr[i], arr[j] = arr[j], arr[i]
			i++
		}
	}
	arr[i], arr[r] = arr[r], arr[i]
	return i
}

func quicksort(arr []int, l, r int) {
	if l >= r { return }
	p := partition(arr, l, r)
	quicksort(arr, l, p-1)
	quicksort(arr, p+1, r)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n int
	fmt.Fscan(reader, &n)
	arr := make([]int, n)
	for i := range arr { fmt.Fscan(reader, &arr[i]) }

	start := time.Now()
	quicksort(arr, 0, n-1)
	ms := float64(time.Since(start).Nanoseconds()) / 1e6

	sb := make([]string, n)
	for i, v := range arr { sb[i] = fmt.Sprintf("%d", v) }
	fmt.Fprintln(writer, strings.Join(sb, " "))
	fmt.Fprintf(os.Stderr, "%.3f\n", ms)
}

File: algorithms/python/quicksort.py
import sys
import time

sys.setrecursionlimit(2000000)

def partition(arr, l, r):
    pivot, i = arr[r], l
    for j in range(l, r):
        if arr[j] <= pivot:
            arr[i], arr[j] = arr[j], arr[i]
            i += 1
    arr[i], arr[r] = arr[r], arr[i]
    return i

def quicksort(arr, l, r):
    if l >= r:
        return
    p = partition(arr, l, r)
    quicksort(arr, l, p - 1)
    quicksort(arr, p + 1, r)

def main():
    data = sys.stdin.read().split()
    n    = int(data[0])
    arr  = list(map(int, data[1:n+1]))

    start = time.perf_counter()
    quicksort(arr, 0, n - 1)
    ms = (time.perf_counter() - start) * 1000

    sys.stdout.write(" ".join(map(str, arr)) + "\n")
    sys.stderr.write(f"{ms:.3f}\n")

if __name__ == "__main__":
    main()